core/src/prove.rs

```diff
@@ -0,0 +1,33 @@
+//! Utilities for proving the results of operations on the trie.
+
+#![cfg(feature = "std")]
+
+use crate::trie::{KeyPath, LeafData, Node};
+
+/// A path to a node in the trie.
+pub struct BitPath {
+    /// The path through the trie. Only the first `len` bits matter.
+    pub path: [u8; 32],
+    /// The length of the path.
+    pub len: u8,
+}
+
+/// A sparse merkle multiproof adapted for NOMT.
+pub struct Proof {
+    /// The leaves and terminators required to prove the values of the proven keys, tagged with
+    /// their depths.
+    ///
+    /// If the `LeafData` is `Some`, this is a leaf. Otherwise, it's a terminal.
+    /// Leaves can be used either to prove the value of a key or to prove that a key doesn't exist
+    /// in the trie by contradiction. Terminals can be used to prove that a key doesn't exist in the
+    /// trie.
+    ///
+    /// These should be sorted by depth (descending) and then lexicographically (ascending) within
+    /// thesame depth.
+    pub relevant_terminals: Vec<(BitPath, Option<LeafData>)>,
+    /// All additional nodes required as part of this proof. These are ordered positionally.
+    pub siblings: Vec<Node>,
```

Comment:
What do you mean by 'positionally'? Is the position related to the 'number rule'?

I see a possible implementation if the siblings are stored in order as needed, and also the terminal nodes are not provided as you described, but they should just be ordered from left to right (if you imagine the tree drawn with all branches going down). The pseudocode would be:

```pseudo

sibiling_index = 0;

for terminal_index in 0..relevant_terminals.len() - 1 {
    
    // check if the current terminal and the next one are sibiling, so last bit 
    // in the sense that both key-path should be of the same lenght and only the last bit
    // of the usefull bit in the key path should differ
    let current_node = relevant_terminals[terminal_index]

    loop {
    
        if differ_last_bit(current_node, relevant_terminals[terminal_index + 1])
            // if nodes are sibiling we can construct the next node
            // the keypath of this new internal is just the previous removed the last one
            internal = hash(current_node, relevant_terminals[terminal_index + 1])
        else if differ_last_bit(current_node, sibiling[sibiling_index]) 
            // if they are not sibiling then you take the one from the proposed sibiling.
            // as before the key path is just the current_node key_path removed the last one
            inernal = hash(current_node, sibiling[sibiling_index])
            sibiling_index += 1
        else 
            // if neither the next terminator or the current sibiling is a real sibiling then 
            // this sub-tree is finished and we can move to the next one
            // the current node will be used as sibiling as soon as all the other sub-tree reach
            // this state
            sibiling.push(internal)
        
    }
    
}

// now apply the same rule to the last terminal using all created sibiling will 
// end up in a roo
loop {

    if sibiling_index >= sibiling.len() 
        // internal is the root
        break
    if differ_last_bit(current_node, sibiling[sibiling_index]) 
        // this should always be true if the my algorithm is correct
        inernal = hash(current_node, sibiling[sibiling_index])
        sibiling_index += 1
}
```

Reply:
> positionally here means "ordered by depth and left-to-right" effectively

Okay! It makes more sense now! But implementing an algorithm that creates or verifies a proof given this structure is not obvious to me, I will think about it!

Reply:
I have some notes. Let me share.
The algorithm is based off of interleaving 3 lists: the terminals, all nodes generated during the verification, and the siblings.

1. instantiate an empty list "implicit_nodes: Vec<(path, node)>"
2. from d (lowest) to 0:
3. pop next terminal with depth d or implicit node with depth d 
4. if the next in implicit or terminals is the sibling, pop that
5. otherwise, set sibling = pop from `siblings`. if `nodes` empty, fail.
6. let new_node = h(node, sibling). push onto end of `implicit_nodes`
7. repeat until all empty
8. check root hash matches.
// complexity: O(logn) in the size of the trie.